src/coreComponents/physicsSolvers/fluidFlow/DissipationCompositionalMultiphaseFVMKernels.hpp

```diff
@@ -0,0 +1,469 @@
+                                           localIndex const (&sei)[2],
+                                           localIndex const connectionIndex,
+                                           localIndex const k_up,
+                                           localIndex const er_up,
+                                           localIndex const esr_up,
+                                           localIndex const ei_up,
+                                           real64 const & potGrad,
+                                           real64 const & phaseFlux,
+                                           real64 const (&dPhaseFlux_dP)[2],
+                                           real64 const (&dPhaseFlux_dC)[2][numComp] )
+    {
+      GEOS_UNUSED_VAR( k_up, potGrad, phaseFlux, dPhaseFlux_dP, dPhaseFlux_dC, er_up, esr_up, ei_up );
+
+      /// Storing dissipation flux and its derivatives locally
+      real64 dissFlux[numComp]{};
+      real64 dDissFlux_dP[numFluxSupportPoints][numComp]{};
+      real64 dDissFlux_dC[numFluxSupportPoints][numComp][numComp]{};
+      real64 fluxPointCoef[numFluxSupportPoints] = {1.0, -1.0}; // for gradients
+
+      real64 viscosityMult[3] = {1.0, 1.0, 1.0}; // for viscosity
+
+      /// Step 1. Calculate the continuation parameter based on the current Newton iteration
+      real64 kappaDBC = 1.0; // default value
+
+      if( m_continuationDBC ) // if continuation is enabled
+      {
+        if( m_curNewton >= 5 )
+        {
+          kappaDBC = m_kappaminDBC;
+        }
+        else
+        {
+          for( int mp = 0; mp < m_curNewton; mp++ ) kappaDBC *= m_contMultiplierDBC;
+          kappaDBC = std::max( kappaDBC, m_kappaminDBC );
+        }
+      }
+      /// Step 2. Collect all contributions
+      real64 poreVolume_n = 0; // Pore volume contribution
+      real64 trans = 0; // Transmissibility contribution
+      for( integer ke = 0; ke < numFluxSupportPoints; ++ke )
+      {
+        trans = std::max( stack.transmissibility[connectionIndex][ke], trans );
+        poreVolume_n += 0.5 *  m_volume[seri[ke]][sesri[ke]][sei[ke]] * m_porosity_n[seri[ke]][sesri[ke]][sei[ke]][0];
+      }
+
+      // potential gradient contribution
+      // pressure
+      real64 pressure_gradient = 0;
+      for( integer ke = 0; ke < numFluxSupportPoints; ++ke )
+        pressure_gradient += fluxPointCoef[ke] * m_pres_n[seri[ke]][sesri[ke]][sei[ke]];
+
+      real64 potential_gradient = LvArray::math::abs( pressure_gradient );
+
+      real64 grad_depth = 0;
+      for( integer ke = 0; ke < numFluxSupportPoints; ++ke )
+      {
+        grad_depth += fluxPointCoef[ke] * m_gravCoef[seri[ke]][sesri[ke]][sei[ke]];
+      }
+
+      // bias towards x and y direction for miscible
+      real64 directional_coef;
+      if( m_miscibleDBC )
+      {
+        directional_coef = 100.0;
+        if( LvArray::math::abs( grad_depth ) != 0 )
+        {
+          if( 1000.f / LvArray::math::abs( grad_depth * grad_depth ) < 100 )
```

Comment:
1. I am not sure I understand the hard-coded threshold criteria here
2. What if gravity is tilted ?
I guess this should be let to the user to specify `m_miscibleBias` as a tensor and this would spare you a `m_miscibleDBC` definition and few if-s

Reply:
this is a bit work in progress to be honest... needs to be generalized and improved

Reply:
@jafranc  added comment in the code. Generalizing to modify from input file left as future work.